rlapy/comps/evd.py

```diff
@@ -0,0 +1,432 @@
+        raise RuntimeError(msg)
+
+    U, lambda_matrix, Vh = la.svd(C)
+    # Full d × d Hermitian eigendecomposition for the smaller matrix C.
+    #Alternatively, U, lambda_matrix = la.eigh(C), but this returns a dense lambda_matrix matrix.
+    r = min(k,d)
+    I = np.argsort(-1*np.abs(lambda_matrix))[range(r)]
+    # indices of r largest components of |λ|
+    U = U[:,I]
+    lambda_matrix = lambda_matrix[I] 
+    V = Q @ U
+    lambda_matrix = np.diag(lambda_matrix)
+    return V, lambda_matrix
+
+def evd2(num_passes, A, k, epsilon, s, rng):
+    """
+    Return the eigen decomposition matrices (V, lambda_matrix),
+    based on a rank-k QB factorization of A.
+    Use a Gaussian sketching matrix and pass over A a total of
+    num_passes times.
+
+    Parameters
+    ----------
+    num_passes : int
+        Total number of passes over A. We require num_passes >= 2.
+        
+    A : Union[ndarray, spmatrix, LinearOperator]
+        Data matrix to approximate. A must be an n × n Hermitian PSD matrix. But we did not check PSD
+
+    k : int
+        Target rank for the approximation of A: 0 < k < min(A.shape).
+        This parameter includes any oversampling. For example, if you
+        want to be near the optimal (Eckhart-Young) error for a rank 20
+        approximation of A, then you might want to set k=25.
+        
+    epsilon = np.nan : NaN by default
+        Target accuracy for the oversampled approximation of A: 0 < epsilon < np.inf.
+        This parameter inherits from the QBDecomposer or RangeFinder class.
+        
+    s  : int
+        Auxiliary parameter for the QBDecomposer or RangeFinder.
+
+    rng : Union[None, int, SeedSequence, BitGenerator, Generator]
+        Determines the numpy Generator object that manages randomness
+        in this function call.
+
+    Returns
+    -------
+    V : ndarray
+        Has shape (A.shape[0], k). Columns are orthonormal.
+
+    lambda_matrix : ndarray
+        Has shape (k,k). lambda_matrix is a diagonal matrix storing the eigenvalues of 
+        the matrix A
+
+    Notes
+    -----
+    We perform (num_passes - 2) steps of subspace iteration, and
+    stabilize subspace iteration by a QR factorization at every step.
+
+    References
+    ----------
+    This algorithm computes V and then sets B = Q.T @ A. Conceptually, we
+    compute Q by adapting Algorithm 5.3 from
+
+        Joel A Tropp, Alp Yurtsever, Madeleine Udell, and Volkan Cevher.
+        "Fixed-rank approximation of a positive-semidefinite matrix from streaming data."
+        Advances in neural information processing systems, 2017.
+        (available at `arXiv <https://arxiv.org/abs/1706.05736>`_).
+    EVDecomposer: psd matrices only, [TYUC17a, Algorithm 3]
+    """
+    if not (A == A.T).all():
+        #Needs to consider conjugate as well if A contains complex numbers.
+        msg = """
+        This rountine evd1 only works for Hermitian matrices.
+        """ 
+        raise RuntimeError(msg)
```

Comment:
Both EVD1 and EVD2 need the matrix to be symmetric / Hermitian. EVD2 needs the matrix to be positive semidefinite, but it isn't possible to check that condition on input.